build/go.bzl

```diff
@@ -0,0 +1,50 @@
+# Copyright 2018 The Kubernetes Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# A wrapper for go_library() which adds .gofiles, .sourcefiles filegroup targets.
+#
+# Usage:
+#   replace lines like:
+#     load("@io_bazel_rule_go//go:def.bzl", "go_library")
+#   with:
+#     load("//build:go.bzl", "go_library")
+#
+#   Now a //foo:go_default_library also comes with //foo:go_default_library.sourcefiles
+
+load("@io_bazel_rules_go//go:def.bzl",
+     "go_test",
+     "go_binary",
+     "go_prefix",
+     _go_library="go_library",
+)
+
+def go_library(name, srcs=None, deps=None, **kw):
+  """Wrap go_library() to also produce a .gofiles and .sourcefiles targets."""
+  # Create the actual library rule
+  _go_library(name=name, srcs=srcs, deps=deps, **kw)
+  # Now also create the .gofiles and .sourcefiles rules
+  srcs = srcs or []
+  deps = deps or []
+  # All go files in this package
+  native.filegroup(
+    name = name+".gofiles",
+    srcs = native.glob(["*.go"]),
+    visibility = ["//visibility:public"],
+  )
+  # All non-generated go files in this package
+  native.filegroup(
+    name = name+".sourcefiles",
+    srcs = native.glob(["*.go"], exclude=["zz_generated.*.go"]),
```

Comment:
the glob worries me a little, since it may not match the srcs of the go_library itself. what's the reason to not use the supplied `srcs` arg and filter out the `zz_generated` files?

Reply:
This is what I tried initially. 

Unfortunately when this macro is evaluated the exact input srcs are not yet known (due to things like [select](https://docs.bazel.build/versions/master/be/functions.html#select)).

Reply:
it looks like this issue might be mitigated if we update to gazelle 0.11, which no longer puts `select()` statements in `srcs`, apparently. https://github.com/bazelbuild/bazel-gazelle/releases/tag/0.11.0

Reply:
I don't see a need to do that now.

Reply:
another take: I wonder if the `.gofiles` and `.sourcefiles` filegroups are even necessary, since the list of source files re already included in `go_library`, exposed through the `GoSource` and `GoArchiveData` providers. (This is how `go_genrule` figures out the list of sources needed, provided a list of `go_library` labels.)

I acknowledge that that makes things a little trickier, since it'd also be necessary to filter out the `zz_generated` files, but it would also mean we wouldn't need the awkward "overload `go_library` requirement either.